hydra-node/src/Hydra/Network/Ouroboros.hs

```diff
@@ -129,34 +130,62 @@ withOuroborosNetwork tracer localHost remoteHosts networkCallback between = do
   let newBroadcastChannel = atomically $ dupTChan bchan
   -- TODO: Factor this out, there should be only one IOManager per process.
   withIOManager $ \iomgr -> do
-    race_ (connect iomgr newBroadcastChannel hydraClient) $
+    -- init a signal mvar used by Network.setPeers to communicate with thread 'connect'
+    signal <- newMVar remoteHosts
+    -- init read signal to avoid blocking on Network.getPeers
+    readSignal <- newMVar remoteHosts
+    race_ (connect iomgr newBroadcastChannel hydraClient signal) $
       race_ (listen iomgr hydraServer) $ do
         between $
           Network
             { broadcast = atomically . writeTChan bchan
+            , getPeers = readMVar readSignal
+            , setPeers = \peers -> do
+                putMVar signal peers
+                modifyMVar readSignal peers
             }
  where
+  modifyMVar signal peers = do
+    _ <- takeMVar signal
+    putMVar signal peers
+
   resolveSockAddr Host{hostname, port} = do
     is <- getAddrInfo (Just defaultHints) (Just $ toString hostname) (Just $ show port)
     case is of
       (info : _) -> pure $ addrAddress info
       _ -> error "getAdrrInfo failed.. do proper error handling"
 
-  connect iomgr newBroadcastChannel app = do
+  -- | only thread connects requires a signal.
+  -- | on signal change, it will release all previous allocated resources
+  -- | and reconnect using the new peers signaled.
+  connect iomgr newBroadcastChannel app signal = do
+    -- take the signaled peers, or wait until there is any.
+    peers <- takeMVar signal
+
     -- REVIEW(SN): move outside to have this information available?
     networkState <- newNetworkMutableState
     -- Using port number 0 to let the operating system pick a random port
     localAddr <- resolveSockAddr localHost{port = 0}
-    remoteAddrs <- forM remoteHosts resolveSockAddr
+
+    remoteAddrs <- forM peers resolveSockAddr
     let sn = socketSnocket iomgr
-    Subscription.ipSubscriptionWorker
-      sn
-      (contramap (WithHost localHost . TraceSubscriptions) tracer)
-      (contramap (WithHost localHost . TraceErrorPolicy) tracer)
-      networkState
-      (subscriptionParams localAddr remoteAddrs)
-      (actualConnect iomgr newBroadcastChannel app)
+    
+    -- run subscription worker until new signal is received
+    -- race_ will break and release allocated resources once we received a new signal.
+    race_ (readMVar signal)
+      ( Subscription.ipSubscriptionWorker
+          sn
+          (contramap (WithHost localHost . TraceSubscriptions) tracer)
+          (contramap (WithHost localHost . TraceErrorPolicy) tracer)
+          networkState
+          (subscriptionParams localAddr remoteAddrs)
+          (actualConnect iomgr newBroadcastChannel app)
+      )
 
+    -- now we can safetly restart the connection
+    -- because connect will wait until new signal is received.
+    connect iomgr newBroadcastChannel app signal
```

Comment:
I'd suggest to use [forever](https://hackage.haskell.org/package/base-4.17.0.0/docs/Control-Monad.html#v:forever) in this situation, which is a bit more idiomatic and avoid potential space-leak (though there's none here).

Reply:
what i want is to run another thread awaiting for a change to execute a reconnection.
i don't think this `forever` does the same, are you sure?

Reply:
I am pretty sure yes.

Reply:
⚠️ by introducing the change of `TVar` and `forever` we have changed the behaviour of thread `connect`,
as now it will restart in case there is a problem with the `subscription worker`.